### PR TITLE
Upgrade LRU-Cache

### PR DESCRIPTION
**Pull request by mattgodbolt**
- rephrase a few LRUs in terms of the new API
- add `okToCache` to the `CompilationResult` to make typing more correct (I _think_)

### CODE REVIEW

File: static/panes/compiler.ts [hunk 1]
```diff
@@ -3329,8 +3333,8 @@ export class Compiler extends MonacoPane<monaco.editor.IStandaloneCodeEditor, Co
         const cacheName = `asm/${instructionSet}/${opcode}`;
         const cached = OpcodeCache.get(cacheName);
         if (cached) {
-            if (cached.found) return cached.data;
-            throw new Error(cached.data);
+            if (cached.found) return cached.data as AssemblyInstructionInfo;
```
Comment (mattgodbolt) [inline static/panes/compiler.ts:3336]: @jeremy-rifkin this is the bit I think I could do better on. Actually we could change the design here - what would you suggest? Maybe get rid of `found` and use the presence or absence of `data` (with a separate `errorString`?)

Comment (jeremy-rifkin) [reply static/panes/compiler.ts:3336]: Good spot, I think that sounds perfect to me!

**Review by jeremy-rifkin — APPROVED** [review]
Looks good to me

File: lib/cache/on-disk.ts [hunk 2]
```diff
@@ -105,6 +105,6 @@ export class OnDiskCache extends BaseCache {
             size: value.length,
         };
         await fs.writeFile(info.path, value);
-        return this.cache.set(key, info);
+        this.cache.set(key, info);
```
[the remark this answers is not in the document]
Comment (mattgodbolt) [reply lib/cache/on-disk.ts:108]: The cache isn't async, no (there's a "fetch" / await "fetch" thing but we don't use it). It's just a dictionary under the hood.

File: types/compilation/compilation.interfaces.ts [hunk 3]
```diff
@@ -34,6 +34,7 @@ import {LLVMOptPipelineOutput} from './llvm-opt-pipeline-output.interfaces.js';
 export type CompilationResult = {
     code: number;
     timedOut: boolean;
+    okToCache?: boolean;
```
Comment (RubenRBS) [inline types/compilation/compilation.interfaces.ts:37]: I think this is right, good catch

**Review by RubenRBS — APPROVED** [review]

File: types/compilation/compilation.interfaces.ts (hunk 3 above)
Comment (mattgodbolt) [reply types/compilation/compilation.interfaces.ts:37]: Thanks for checking, that was the most worrying bit :)

**Pull request merged**